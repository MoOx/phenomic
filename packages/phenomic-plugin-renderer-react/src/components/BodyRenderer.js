/* eslint-disable react/no-multi-comp */
// because it considers render() as a react compo
/* eslint-disable react/no-unused-prop-types */
// because it's a bit dumb

import React from "react"
// @todo add Phenomic Link
// import Link from "./Link.js"

type ComponentType = string | Class<React.Component<*, *, *>> | Function

type OptionsType = {
  components: { [key: string]: ComponentType },
  DefaultComponent: ComponentType,
}

type ItemType =
  | string
  | {
      // tag
      t?: string,
      // props
      p?: Object,
      // children
      // @todo use this type
      // https://github.com/yannickcr/eslint-plugin-react/pull/1138
      // note that eslint-disable is not working as it remove, after being found
      // errors and warnings
      // c: ItemType | Array<ItemType>,
      c: Object,
    }

type PropsType = {
  children: string | ItemType,
  options: OptionsType,
}

const defaultOptions: OptionsType = {
  components: {
    // a: Link
  },
  DefaultComponent: "div",
}

const render = (item: ItemType, options: OptionsType) => {
  if (!item) {
    return
  }
  if (typeof item === "object") {
    const {
      t: Tag = options.DefaultComponent,
      p: props = {},
      c: children,
    } = item
    return (
      <Tag {...props}>
        {Array.isArray(children)
          ? children.map((child: ItemType) => render(child, options))
          : render(children, options)}
      </Tag>
    )
  }
  return item
}

const BodyRenderer = (props: PropsType) => {
  if (typeof props.children === "string") {
    return <div dangerouslySetInnerHTML={{ __html: props.children }} />
  }
  return render(props.children, { ...defaultOptions, ...(props.options || {}) })
}

export default BodyRenderer
