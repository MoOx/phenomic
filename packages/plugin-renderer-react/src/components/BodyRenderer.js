/* eslint-disable react/no-multi-comp */
// because it considers render() as a react compo
/* eslint-disable react/no-unused-prop-types */
// because it's a bit dumb

import React from "react";

import Link from "./Link.js";

type ComponentType = string | Class<React.Component<*, *, *>> | Function;

type OptionsType = {
  components?: { [key: string]: ComponentType },
  DefaultComponent?: ComponentType
};

type ItemType =
  | string
  | {
      // tag
      t?: string,
      // props
      p?: Object,
      // children
      // @todo use this type
      // https://github.com/yannickcr/eslint-plugin-react/pull/1138
      // note that eslint-disable is not working as it remove, after being found
      // errors and warnings
      // c: ItemType | Array<ItemType>,
      c: Object
    };

type PropsType = OptionsType & {
  children?: string | ItemType
};

const defaultOptions: OptionsType = {
  components: {
    a: Link
  }
};

const render = (item: ItemType, options: OptionsType, key: ?any) => {
  if (!item) {
    return;
  }
  if (typeof item === "string") {
    return item;
  }
  const { p: props = {}, c: children } = item;
  const Tag =
    (options.components && item.t && options.components[item.t]) ||
    item.t ||
    options.DefaultComponent ||
    "div";
  return (
    <Tag {...props} key={key}>
      {Array.isArray(children)
        ? children.map((child: ItemType, key) => render(child, options, key))
        : render(children, options)}
    </Tag>
  );
};

const BodyRenderer = ({ children, ...props }: PropsType) => {
  if (typeof children === "undefined") {
    console.error(
      "@phenomic/plugin-renderer-react: BodyRenderer expects at least a child"
    );
    return null;
  }
  if (typeof children === "string") {
    return <div dangerouslySetInnerHTML={{ __html: children }} />;
  }
  const r = render(children, {
    ...defaultOptions,
    ...(props || {}),
    // force to mix components, as default one (Link) is crucial
    components: {
      ...defaultOptions.components,
      ...(props.components || {})
    }
  });
  return typeof r === "string"
    ? <props.DefaultComponent>{r}</props.DefaultComponent>
    : r;
};

export default BodyRenderer;
