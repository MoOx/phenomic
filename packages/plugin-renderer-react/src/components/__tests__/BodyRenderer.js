/* eslint-disable react/no-multi-comp */

import React from "react";
import { render, renderJSX } from "jsx-test-helpers";

import BodyRenderer from "../BodyRenderer";

test("BodyRenderer tells you when pass undefined", () => {
  const spy = jest.spyOn(console, "error");
  expect(render(<BodyRenderer />)).toMatchSnapshot();
  expect(spy).toHaveBeenCalled();
});

test("BodyRenderer works with raw html", () => {
  expect(
    renderJSX(<BodyRenderer>{"<h1>Test</h1"}</BodyRenderer>)
  ).toMatchSnapshot();
});

test("BodyRenderer works with phenomic json body", () => {
  expect(
    renderJSX(
      <BodyRenderer>
        {{
          t: "p",
          p: { className: "test" },
          c: [
            {
              t: "a",
              p: { href: "http://test" },
              c: "Link"
            }
          ]
        }}
      </BodyRenderer>
    )
  ).toMatchSnapshot();
});

test("BodyRenderer accepts a map for components", () => {
  const map = {
    p: function View(props: Object) {
      return <div>{props.children}</div>;
    },
    a: function Link(props: Object) {
      return <a href={props.href} className="auto">{props.children}</a>;
    }
  };
  expect(
    renderJSX(
      <BodyRenderer components={map}>
        {{
          t: "p",
          p: { className: "test" },
          c: [
            {
              t: "a",
              p: { href: "http://test" },
              c: "Link"
            }
          ]
        }}
      </BodyRenderer>
    )
  ).toMatchSnapshot();
});
