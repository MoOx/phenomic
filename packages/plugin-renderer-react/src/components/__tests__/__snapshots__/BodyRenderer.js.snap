// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BodyRenderer accepts a map for components 1`] = `
"<View className=\\"test\\">
  <Link href=\\"http://test\\">
    Link
  </Link>
</View>"
`;

exports[`BodyRenderer tells you when pass undefined 1`] = `null`;

exports[`BodyRenderer works with phenomic json body 1`] = `
"<p className=\\"test\\">
  <Link href=\\"http://test\\">
    Link
  </Link>
</p>"
`;

exports[`BodyRenderer works with raw html 1`] = `
"<div
  dangerouslySetInnerHTML={{
    __html: '<h1>Test</h1'
  }}
 />"
`;
